Improve Fee Estimation Logic Using EIP-1559

ArturDolzan opened the pull request:
### Description

This PR improves the internal logic of the useEstimateFees hook to provide more consistent EIP-1559 fee estimations.. The total fee is estimated by computing **gasUnits * maxFeePerGas, where maxFeePerGas = baseFee * 2 + priorityFee**.

As a result of these improvements, I think the overestimation factor could be reduced from 1.5 to around 1.2. This would be a very small refinement though.

References: [EIP-1559 - Cheatsheet for Implementers](https://hackmd.io/4YVYKxxvRZGDto7aq7rVkg?view)

### Screenshots

#### Comparatives

**Sepolia** (L1 to L2)
TX: https://sepolia.etherscan.io/tx/0x1175eac2f09cf6f90b275443d9816a822e5b05fac613f48c3db78437fa121f84
<img width="824" alt="Captura de Tela 2025-06-02 às 13 53 23" src="https://github.com/user-attachments/assets/fb3f19bc-1adb-482c-a47a-74ec77ae66e4" />


**Mainnet** (L1 to L2)
TX: https://etherscan.io/tx/0x4665951d3979b0dd088856bcc14e941b4c3348fdb367c881fb70e751606908c6
<img width="817" alt="Captura de Tela 2025-06-02 às 14 14 14" src="https://github.com/user-attachments/assets/f7653088-9712-4082-aa15-baee518b31b8" />

**Mainnet summary** (L1 to L2)
<img width="582" alt="Captura de Tela 2025-06-02 às 14 15 45" src="https://github.com/user-attachments/assets/e1f5260d-785d-474f-939a-8658d2f78d32" />

**Sepolia** (L2 to L1)
<img width="547" alt="Captura de Tela 2025-06-02 às 20 47 07" src="https://github.com/user-attachments/assets/c81840c4-7bbd-452f-8594-efa4ee3cdbe0" />

**Mainnet** (L2 to L1) 
Note: Similar to MM estimations, which was also far above the actual fee.
<img width="600" alt="Captura de Tela 2025-06-02 às 20 55 59" src="https://github.com/user-attachments/assets/d8fddbd9-1475-4661-8c85-ce734b0a65db" />

Note the last comparison 

### Related issue(s)

Closes #1124 
Fixes #1124

### Checklist

- [x] Manual testing passed.
- [ ] Automated tests added, or N/A.
- [ ] Documentation updated, or N/A.
- [ ] Environment variables set in CI, or N/A.

gndelia commented on portal/hooks/useEstimateFees.ts:
```diff
@@ -6,6 +6,10 @@ import { useIsConnectedToExpectedNetwork } from './useIsConnectedToExpectedNetwo
+ * Returns the average as a stringified number.
+ */
 const mean = function (rewards: bigint[] = []) {
   if (rewards.length === 0) {
     return '0'
```
Given this function is only used locally, this line will never run, as in line 71 where this function is called, we're ensuring `rewards` are always defined and with `.length > 0`. The same applies to the default ` = []`

ArturDolzan commented on portal/hooks/useEstimateFees.ts:
```diff
@@ -45,17 +62,28 @@ export const useEstimateFees = function ({
 
-  const gasUnits = props.gasUnits ?? BigInt(0)
+  // Use the 30th percentile reward as a proxy for maxPriorityFeePerGas
+  const prioritySamples = feeHistory?.reward.map(r => r.at(-1)) ?? []
```
Not needed.

ArturDolzan commented:
The main change was removing the custom mean function (which manually averaged priority fees from feeHistory) and replacing it with the official useEstimateMaxPriorityFeePerGas hook from wagmi, which is more reliable. It was also added a fallbackPriorityFeeByChain map so we can handle edge cases like hemiSepolia, where the priority fee needs to be much lower. 

It was also updated the PR description with comparison tables showing our estimate, MM estimate, and the actual fee paid across scenarios like L1 to L2 and L2 to L1.

gndelia commented on portal/hooks/useEstimateFees.ts:
```diff
@@ -33,35 +62,58 @@ export const useEstimateFees = function ({
   return {
-    fees,
+    fees: BigInt(totalFeeEstimate.toFixed(0)),
     isError: isGasUnitsError,
```
Perhaps here we need to add the errors of `useEstimateMaxPriorityFeePerGas` and `useFeeHistory` with an `||`

ArturDolzan replied:
Done. Thanks!

gndelia commented:
I run some tests comparing Testnet estimated fees vs Actual paid fees:

#### Deposit (Sepolia)

![image](https://github.com/user-attachments/assets/9b93b73a-f470-4412-8de9-6b8876c67242)

Estimation: `0.000939` <-> Actual fees paid: `0.0008` (See [tx](https://sepolia.etherscan.io/tx/0x3bbd8f80e304e9c3bb15dad22b3d881fd4d3be2776bc485465a5d2fbc7c1cc73))

#### Initiate Withdraw (Hemi Sepolia)

![image](https://github.com/user-attachments/assets/4de41079-392d-4aa8-9a85-ef7a27a7af0c)

Initiate Withdraw (Sepolia)
Estimation: `0.0000002118` <-> Actual fees paid: `0.000000136695663769` (See [tx](https://testnet.explorer.hemi.xyz/tx/0xa7f9ca5f12269f23250cec6fd1f8680067e1f49b3f0477447e254f0b7160d7c0))


#### Prove Withdrawal (Sepolia)

![image](https://github.com/user-attachments/assets/55271b13-29db-42e7-a86f-65559b7d5f26)

Prove Withdrawal (Sepolia)
Estimation: `0.000491` <-> Actual fees paid: `0.0004802`  (See [tx](https://sepolia.etherscan.io/tx/0xd8024855820dbe0d6a0a167d4016252fc0e91064073e6b88fce4769885279efd))

#### Claim Withdrawal (Sepolia)

![image](https://github.com/user-attachments/assets/869551c4-e4c4-46ee-8145-3bc4a43c4877)

Estimation: `0.000661` <-> Actual fees paid `0.000299` (See [tx](https://sepolia.etherscan.io/tx/0xe3c2fe287e62573a7ba829880118f01d83f96e01493a266c8b05ec745a174487))


----

So far looking good, I will now test Mainnet

gndelia commented:
FYI fees are high on mainnet, so I may test it later at night when fees are lower

gndelia commented:
I initiated a Withdrawal in Hemi mainnet

![image](https://github.com/user-attachments/assets/916f273a-5bcb-4ea2-8a1f-b433757e3661)

The estimation had 10 zeros, however, the actual fees paid were `0.000008509073286416` - See [tx](https://explorer.hemi.xyz/tx/0x67a26196da58da366a93de9c0cca6f1659a134823192eaddcd554de255595b8f). It seems there are still something we're not quite figuring ok 🤔

gndelia commented:
#### Initiate Withdraw (Hemi)

![image](https://github.com/user-attachments/assets/c50e92c4-547b-4617-9cb2-d60dc7fad0b5)

Estimated fees were `0.0001920` <-> Actual paid fees `0.000007982865684184` (See [tx](https://explorer.hemi.xyz/tx/0xca1b5f2538887b6dc5eb3148879a5eefcaa06dbf92d67fc119fa5c3e8798a9c99))


#### Prove Withdrawal (Ethereum)

<img width="455" alt="image" src="https://github.com/user-attachments/assets/b7d0f185-27f1-4d3c-89e4-173f897bd990" />

Estimated fees were `` <-> Actual paid fees `0.00106126049586128` (See [tx](https://etherscan.io/tx/0xeda0bf619f5622ae637c382662822d504f83f9aa6c51549f0d23853183d392bb))

gndelia commented:
I have no withdrawal to Claim yet on Mainnet. My hunch tells me that Mainnet and Sepolia estimations are fairly accurate, and Hemi are a bit less so (although they have improved from the current estimations)

ArturDolzan commented:
> #### Initiate Withdraw (Hemi)
> ![image](https://private-user-images.githubusercontent.com/352474/451050633-c50e92c4-547b-4617-9cb2-d60dc7fad0b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkwMzA3NzMsIm5iZiI6MTc0OTAzMDQ3MywicGF0aCI6Ii8zNTI0NzQvNDUxMDUwNjMzLWM1MGU5MmM0LTU0N2ItNDYxNy05Y2IyLWQ2MGRjN2ZhZDBiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYwNFQwOTQ3NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zN2M2N2E4MDRmMDY5MDc0ODdlOWVjY2IxM2U4NThhMzNjNzI1ZDIyNTg5OGJhNWM2ZjI5MGQ3YmI0OTZkZDJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Zj4nlJZG4nFPVH5gUjCLyjZBp-TldnAepdDOkdN6YuA)
> 
> Estimated fees were `0.0001920` <-> Actual paid fees `0.000007982865684184` (See [tx](https://explorer.hemi.xyz/tx/0xca1b5f2538887b6dc5eb3148879a5eefcaa06dbf92d67fc119fa5c3e8798a9c99))
> 
> #### Prove Withdrawal (Ethereum)
> <img alt="image" width="455" src="https://private-user-images.githubusercontent.com/352474/451051326-b7d0f185-27f1-4d3c-89e4-173f897bd990.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkwMzA3NzMsIm5iZiI6MTc0OTAzMDQ3MywicGF0aCI6Ii8zNTI0NzQvNDUxMDUxMzI2LWI3ZDBmMTg1LTI3ZjEtNGQzYy04OWU0LTE3M2Y4OTdiZDk5MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYwNFQwOTQ3NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMzcwNDk5NzkzZjZlNjEzM2U0NWMxOTUwZTExZGU0ZTQzN2NkM2RjYWU1ZjUyZjQzZWZiMGRlYjJjNmU4YWUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2ze8TKPzHL350MD8yZ0ijVjY86pwi9bmHEalY7rIISA">
> Estimated fees were `0.00267` <-> Actual paid fees `0.00106126049586128` (See [tx](https://etherscan.io/tx/0xeda0bf619f5622ae637c382662822d504f83f9aa6c51549f0d23853183d392bb))

@gndelia yes, this is very similar to what I got during my tests. The estimate is closer to MetaMask, but the actual fee paid is way lower. I think we can improve the accuracy with some tuning (like I did for Hemi Sepolia), since gas costs on L2 networks are usually very low. I can definitely tackle this in future PRs.

gndelia reviewed (approved):
let's get this merged and we'll monitor it for future updates
